src/torch_ucc.cpp

```diff
@@ -220,17 +226,92 @@ void ProcessGroupUCC::ProgressEntry::finalize(std::exception_ptr eptr) {
   }
 }
 
+ucs_status_t torch_ucc_timeout_am_cb(
+    void *arg,
+    const void *header,
+    size_t header_length,
+    void *data,
+    size_t length,
+    const ucp_am_recv_param_t *param) {
+  torch_ucc_timeout_desc_t *dsc = (torch_ucc_timeout_desc_t*)header;
+  CommPG *comm = (CommPG*)arg;
+  torch_ucc_rank_state_t * state = new torch_ucc_rank_state_t;
+  ucp_request_param_t params;
+  ucp_tag_t ucp_tag;
+
+#ifdef USE_CUDA
+  if (comm->cuda_device_index >= 0) {
+    CUdevice device;
+    CUresult st;
+    unsigned int flags;
+    int active;
+    st = cuDeviceGet(&device, comm->cuda_device_index);
+    if (st != CUDA_SUCCESS) {
+      *state = TORCH_UCC_RANK_STATE_DEVICE_ERROR;
+      goto send_response;
+    }
+    st = cuDevicePrimaryCtxGetState(device, &flags, &active);
+    if (st != CUDA_SUCCESS || !active) {
+      *state = TORCH_UCC_RANK_STATE_DEVICE_ERROR;
+      goto send_response;
+    }
+  }
+#endif
+
+  *state = TORCH_UCC_RANK_STATE_COLLECTIVE_NOT_POSTED;
+  if (comm->seq_num > dsc->seq_num) {
+    *state = TORCH_UCC_RANK_STATE_COLLECTIVE_DONE;
+    for (auto& entry : comm->progress_queue) {
+      if (entry->seq_num_ == dsc->seq_num) {
+        switch (entry->request_->status)
+        {
+        case UCC_ERR_TIMED_OUT:
+          *state = TORCH_UCC_RANK_STATE_COLLECTIVE_TIMEOUT;
+          break;
+        default:
+          *state = TORCH_UCC_RANK_STATE_COLLECTIVE_INPROGRESS;
+        }
+        break;
+      }
+    }
+  }
+
+send_response:
+  TORCH_UCX_MAKE_OOB_SEND_TAG(ucp_tag, 0, dsc->rank, dec->comm_id);
+  params.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+      UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FIELD_USER_DATA;
+  params.datatype = ucp_dt_make_contig(sizeof(torch_ucc_rank_state_t));
+  params.user_data = (void*)state;
+  params.cb.send = [](void* request, ucs_status_t status, void* user_data) {
+    torch_ucc_rank_state_t * state = (torch_ucc_rank_state_t*)user_data;
```

Comment:
nit: 
is this required? we don't use `state` at all here since we do `delete user_data` next line. or do you intent to do `delete state`?